lib/ftu/signuplogin.dart

```diff
@@ -0,0 +1,253 @@
+
+  String _email, _password, _errorMessage;
+  bool _isLoginForm, _isLoading;
+
+  // Check if form is valid before perform login or signup
+  bool validateAndSave() {
+    final form = _formKey.currentState;
+    if (form.validate()) {
+      form.save();
+      return true;
+    }
+    return false;
+  }
+
+  // Perform login or signup
+  void validateAndSubmit() async {
+    setState(() {
+      _errorMessage = "";
+      _isLoading = true;
+    });
+    if (validateAndSave()) {
+      String userId = "";
+      try {
+        if (_isLoginForm) {
+          userId = await widget.auth.signIn(_email, _password);
+          print('Signed in: $userId');
+        } else {
+          userId = await widget.auth.signUp(_email, _password);
+          widget.auth.sendEmailVerification();
+          _showVerifyEmailSentDialog();
+          print('Signed up user: $userId');
+        }
+        setState(() {
+          _isLoading = false;
+        });
+
+        if (userId.length > 0 && userId != null && _isLoginForm) {
+          widget.loginCallback();
+        }
+        // Avoid catches without on clauses.
+      } catch (e) { 
+        print('Error: $e');
+        setState(() {
+          _isLoading = false;
+          _errorMessage = e.message as String;
+          // A value of type 'dynamic' can't be assigned to a variable of type 'String'. 
+          // Try changing the type of the variable, or casting the right-hand type to 'String'.
+          _formKey.currentState.reset();
+        });
+      }
+    }
+  }
+
+  @override
+  void initState() {
+    _errorMessage = "";
+    _isLoading = false;
+    _isLoginForm = true;
+    super.initState();
+  }
+
+  void resetForm() {
+    _formKey.currentState.reset();
+    _errorMessage = "";
+  }
+
+  void toggleFormMode() {
+    resetForm();
+    setState(() {
+      _isLoginForm = !_isLoginForm;
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+        appBar: AppBar(
+          title: Text(welcomeMessage),
+        ),
+        body: Stack(
+          children: <Widget>[
+            _showForm(),
+            _showCircularProgress(),
+          ],
+        ));
+  }
+
+  Widget _showCircularProgress() {
+    if (_isLoading) {
+      return Center(child: CircularProgressIndicator());
+    }
+    return Container(
+      height: 0.0,
+      width: 0.0,
+    );
+  }
+
+ void _showVerifyEmailSentDialog() {
+   showDialog(
+     context: context,
+     builder: (BuildContext context) {
+       // Return object of type Dialog
+       return AlertDialog(
+         title: Text(vertifyaccountTitle),
+         content: Text(vertifyaccountMessage),
+         actions: <Widget>[
+           FlatButton(
+             child: Text(dissmissMessage),
+             onPressed: () {
+               toggleFormMode();
+               Navigator.of(context).pop();
+             },
+           ),
+         ],
+       );
+     },
+   );
+ }
+
+  Widget _showForm() {
```

Comment:
We can abstract this widget and all those below it into a file called `form.dart`.